cpp/src/parquet/encryption/file_key_unwrapper.cc

```diff
@@ -30,33 +30,37 @@ FileKeyUnwrapper::FileKeyUnwrapper(
     const std::string& file_path,
     const std::shared_ptr<::arrow::fs::FileSystem>& file_system)
     : FileKeyUnwrapper(std::move(key_toolkit), /*key_toolkit=*/nullptr,
-                       kms_connection_config, cache_lifetime_seconds, file_path,
-                       file_system) {}
+                       kms_connection_config, cache_lifetime_seconds,
+                       /*key_material_store=*/nullptr, file_path, file_system) {}
 
 FileKeyUnwrapper::FileKeyUnwrapper(
     KeyToolkit* key_toolkit, const KmsConnectionConfig& kms_connection_config,
     double cache_lifetime_seconds, const std::string& file_path,
     const std::shared_ptr<::arrow::fs::FileSystem>& file_system)
     : FileKeyUnwrapper(/*key_toolkit_owner=*/nullptr, key_toolkit, kms_connection_config,
-                       cache_lifetime_seconds, file_path, file_system) {}
+                       cache_lifetime_seconds, /*key_material_store=*/nullptr, file_path,
+                       file_system) {}
 
 FileKeyUnwrapper::FileKeyUnwrapper(
     KeyToolkit* key_toolkit, const KmsConnectionConfig& kms_connection_config,
     double cache_lifetime_seconds,
     std::shared_ptr<FileKeyMaterialStore> key_material_store)
     : FileKeyUnwrapper(/*key_toolkit_owner=*/nullptr, key_toolkit, kms_connection_config,
-                       cache_lifetime_seconds, /*file_path=*/"",
+                       cache_lifetime_seconds, std::move(key_material_store),
```

Comment:
@adamreeve I found previously this `key_material_store` is not used. Is this expected?

Reply:
It seems that FileKeyMaterialStore is only required when master key rotation is used. Otherwise it is null and created internally based on file path and file system.

Reply:
It would be good for @adamreeve to double check if this is required. If not, we probably should remove the parameter to reduce confusion.

Reply:
This is unexpected. It looks like during the process of working on #34181 I refactored things so that this parameter was no longer used but didn't realise this. The `FileKeyUnwrapper` created in `KeyToolkit::RotateMasterKeys` is only ever used to call `GetDataEncryptionKey`, which takes a `KeyMaterial` parameter so bypasses the `key_material_store_` completely, and there's no reason it should need to be passed in to the constructor.

I also added an incorrect comment about this constructor overload in `file_key_unwrapper.h`, saying, "This is useful for key rotation to allow accessing the key material store after it is used." I guess I got myself confused with the use of the `temp_key_material_store` passed to the `FileKeyWrapper`.

It seems to me that this overload can be removed completely, and `KeyToolkit::RotateMasterKeys` can use the other constructor. The `file_path` and `filesystem` could be provided but won't be used.

Reply:
Oh actually it looks like the constructor parameter was being used prior to #40329, I didn't notice an extra commit was added to that PR to refactor the constructor parameters making it no longer used (https://github.com/apache/arrow/pull/40329/commits/7547c79b05e0657f6cd3961d01226c4fc84053af). But even though it was being used to set `key_material_store_` before, that was never actually used and the extra constructor overload wasn't needed.

Reply:
So my edit is correct? Or I should removing this?

Reply:
Yeah I think it depends on whether removing the overload is likely to affect any users. Your change is correct, but @pitrou's change not breaking anything exposed that the extra overload wasn't really needed by `KeyToolkit::RotateMasterKeys`. I guess it doesn't hurt to keep it as is, but it might be nice to remove the overload to simplify things.

Reply:
I agree with the current change to fix the missing argument instead of removing a public function.

Reply:
Can you open an issue for exploring whether the overload can be removed? In the meantime we can merge this PR.

Reply:
Sure